docs/advanced-config.rst
============
Usage Manual
============

The SimFleet platform employs JSON configuration files to define simulation scenarios. Once a configuration file has been
defined, its simulation can be launched by running the application from the console. SimFleet can be executed in two modes:
a **command-line interface (CLI)** and a **web-based graphical interface (GUI)**. You may execute simulations purely through the command
line or use the simpler and more intuitive graphical interface.

In this section, the structure and usage of SimFleet configuration files is described. First, we focus on the general structure.
Then, we present the currently included transportation modes and describe how to setup a simulation using its agents. Finally,
we indicate how to use **command-line interface (CLI)** to launch simulation scenarios with different execution options.


The Configuration file's Structure
==================================

The configuration file of a SimFleet simulation fully characterizes the simulation scenario including the information of every agent.
The file follows a JSON structure, presenting one list per type of agent in the simulation, as well as some general configuration entries.
In turn, the list of each agent type contains individual agent definitions: one dictionary per agent, describing its necessary attributes.

Following, we present a description of each basic field present in default configuration files, splitting the information in sections.

**Agent settings**: Listings of agent definitions, split by specific agent types. Any combination of agents may be included in a simulation scenario,
although their interaction will be dependent on their strategic behaviour.

+--------------------------------------------------------------------------------------+
|  Agent definitions & settings                                                        |
+-------------+------------------------------------------------------------------------+
|  Field      |  Description                                                           |
+=============+========================================================================+
| fleets      |   List of FleetManager agents                                          |
+-------------+------------------------------------------------------------------------+
| transports  |   List of Transport agents                                             |
+-------------+------------------------------------------------------------------------+
| customers   |   List of Customer agents                                              |
+-------------+------------------------------------------------------------------------+
| stations    |   List of Station (infrastructure) agents                              |
+-------------+------------------------------------------------------------------------+
| stops       |   List of BusStop agents (for urban bus simulations)                   |
+-------------+------------------------------------------------------------------------+
| *lines*     |   List of bus lines (for urban bus simulation)                         |
+-------------+------------------------------------------------------------------------+
| vehicles    |   List of autonomous Vehicle agents                                    |
+-------------+------------------------------------------------------------------------+

.. note::
    The field *lines* does not define agents. It is an auxiliary configuration field for bus transportation simulations.

**Default Strategies**: The behaviour of an agent during the simulation is governed by its so-called Strategy.
The configuration file allows for the definition of a specific strategy for each of the introduced agents. However, in most cases,
all agents of the same type will behave in the same manner. For that, the configuration file includes the following fields.

+--------------------------------------------------------------------------------------------------+
|  Default strategic behaviours                                                                    |
+-----------------------+--------------------------------------------------------------------------+
|  Field                |  Description                                                             |
+=======================+==========================================================================+
| transport_strategy    |   The strategic behaviour used by Transport agents                       |
+-----------------------+--------------------------------------------------------------------------+
| customer_strategy     |   The strategic behaviour used by Customer agents                        |
+-----------------------+--------------------------------------------------------------------------+
| fleetmanager_strategy |   The strategic behaviour used by Fleet manager                          |
+-----------------------+--------------------------------------------------------------------------+
| station_strategy      |   The strategic behaviour used by Station agents                         |
+-----------------------+--------------------------------------------------------------------------+
| vehicle_strategy      |   The strategic behaviour used by Vehicle agents                         |
+-----------------------+--------------------------------------------------------------------------+
| directory_strategy    |   The strategic behaviour used by the Directory agent                    |
+-----------------------+--------------------------------------------------------------------------+

**Simulation settings**: General settings for the localization of a scenario and the duration of its simulation.

+---------------------------------------------------------------------------------------------+
|  Simulation settings                                                                        |
+------------------+--------------------------------------------------------------------------+
|  Field           |  Description                                                             |
+==================+==========================================================================+
| simulation_name  |   Name of the simulation defined by this configuration file              |
+------------------+--------------------------------------------------------------------------+
| max_time         |   Maximum time (in seconds) for which the simulation will run            |
+------------------+--------------------------------------------------------------------------+
| coords           |   The initial geographic coordinates for the simulation map              |
+------------------+--------------------------------------------------------------------------+
| zoom             |   The initial zoom level of the simulation map                           |
+------------------+--------------------------------------------------------------------------+

.. note::
    The **coords** field can use the name of a city, town, neighbourhood or a specific coordinate, e.g. ‘Valencia’ or [39.4697065, -0.3763353]. This reference point centres the simulation on the map.
    In addition, the **zoom** field controls the scale of the bounding box for the random creation of positions of an agent on the map.

**Metrics and Server settings:** Among the remaining fields, we highlight **mobility_metrics**, which accepts a path to the file that defines the metrics to be calculated at the end of a simulation.
The rest refer to parameters necessary for the communication of SimFleet agents with the routing server and the XMPP server.
We recommend keeping their default values, which appear in the example configuration files.

+--------------------------------------------------------------------------------------------------------------+
|  Metrics, Credentials and Network Settings                                                                   |
+-----------------------+--------------------------------------------------------------------------------------+
|  Field                |  Description                                                                         |
+=======================+======================================================================================+
| mobility_metrics      |   Custom class for the computation of simulation metrics at the end of the execution |
+-----------------------+--------------------------------------------------------------------------------------+
| fleetmanager_name     |   Name for registering the fleet manager agent in the XMPP server                    |
+-----------------------+--------------------------------------------------------------------------------------+
| fleetmanager_password |   Password for registering the fleet manager agent in the XMPP server                |
+-----------------------+--------------------------------------------------------------------------------------+
| directory_name        |   Name for registering the directory agent in the XMPP server                        |
+-----------------------+--------------------------------------------------------------------------------------+
| directory_password    |   Password for registering the directory agent in the XMPP server                    |
+-----------------------+--------------------------------------------------------------------------------------+
| route_host            |   URL of the OSRM routing service used for calculating agent movement                |
+-----------------------+--------------------------------------------------------------------------------------+
| host                  |   The XMPP host address where the simulation platform is running                     |
+-----------------------+--------------------------------------------------------------------------------------+
| xmpp_port             |   Port for XMPP communication                                                        |
+-----------------------+--------------------------------------------------------------------------------------+
| http_port             |   Port for the HTTP server used for the simulator's GUI                              |
+-----------------------+--------------------------------------------------------------------------------------+
| http_ip               |   IP address for the HTTP server used for the simulator's GUI                        |
+-----------------------+--------------------------------------------------------------------------------------+

This structure provides a flexible framework for creating different scenarios. You may find an empty configuration file below,
to be used as a template for the development of custom simulation scenarios.

.. code-block:: json

    {
    "fleets": [],
    "transports": [],
    "customers": [],
    "stations": [],
    "stops": [],
    "lines": [],
    "vehicles": [],
    "simulation_name": "my_city",
    "max_time": 1000,
    "coords": "Valencia",
    "zoom": 12,
    "transport_strategy": "simfleet.module.file.TransportBehaviourClass",
    "customer_strategy": "simfleet.module.file.CustomerBehaviourClass",
    "fleetmanager_strategy": "simfleet.module.file.FleetManagerBehaviourClass",
    "directory_strategy": "simfleet.module.file.DirectoryBehaviourClass",
    "station_strategy": "simfleet.module.file.StationBehaviourClass",
    "vehicle_strategy": "simfleet.module.file.VehicleBehaviourClass",
    "mobility_metrics": "simfleet.module.file.MyMetricsClass",
    "fleetmanager_name": "fleetmanager",
    "fleetmanager_password": "fleetmanager_passwd",
    "route_host": "http://router.project-osrm.org/",
    "directory_name": "directory",
    "directory_password": "directory_passwd",
    "host": "localhost",
    "xmpp_port": 5222,
    "http_port": 9000,
    "http_ip": "localhost"
    }

Transportation simulation modes
===============================

SimFleet is designed to give its users the tools to easily setup and execute complex transportation scenarios. In addition,
users may extend the provided agents to create new versions that adapt to their needs. However, the platform currently
provides two predefined transportation modes: Taxi/Electric Taxi service, and Urban Bus. Following, the agents employed in
each of the modes are briefly described, together with the necessary attributes to define them in a configuration file.

Taxi service simulation
-----------------------

This transportation mode represents a taxi service coordinated by a centralised manager. Customers of the service send
travel requests to the manager who, in turn, broadcasts them to all available transports in its fleet. Upon the reception
of a customer request, taxi agents may choose to serve such the issuing customer, which emcompases picking them up at their
current position and driving them to their destination. The scenario features three agents: A **FleetManager Agent**,
the **Taxi Agents**, and the **TaxiCustomer Agents**.


Taxi Service Agent description
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* **TaxiCustomer Agents**

    TaxiCustomer agents represent people that need to go from one location of the city (their "current location") to
    another (their "destination").
    For doing so, each customer requests a single transport service and, once it is delivered to its destination,
    it ends its execution.

* **Taxi Agents**

    The Taxi agents represent vehicles which can transport TaxiCustomer agents from their current positions to their respective
    destinations. Taxis spawn available in given locations and react to customer requests received from their fleet manager.

* **FleetManager Agent**

    The FleetManager agent is responsible for putting in contact the TaxiCustomer agents that need a transport service, and the Taxi
    agents that may be available to offer these services. In short, the FleetManager Agent acts like a transport call center, accepting
    the incoming requests from customers and forwarding them to the (appropriate) taxis.
    In order to do so, the FleetManager features a registration protocol that allows Taxi agents to subscribe to their manager.
    This is process is automatically done when a Taxi agent starts its execution.


.. In the context of SimFleet, a "transport service" involves the following steps:

    .. The Taxi moves from its current position to the TaxiCustomer's location to pick them up.
    .. The Taxi transports the TaxiCustomer to their destination.


Taxi Service Configuration file
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Following, the necessary configuration file fields to define the taxi service agents are shown. These include a list of
taxi customers, taxi transports and the fleet manager.

A TaxiCustomer agent is defined with the following fields:

+--------------------------------------------------------------------------------------+
|  Taxi Customer                                                                       |
+-------------+------------------------------------------------------------------------+
|  Field      |  Description                                                           |
+=============+========================================================================+
| class       |   Custom agent file in the format ``module.file.Class``                |
+-------------+------------------------------------------------------------------------+
| position    |   Initial coordinates of the customer (optional)                       |
+-------------+------------------------------------------------------------------------+
| destination |   Destination coordinates of the customer (optional)                   |
+-------------+------------------------------------------------------------------------+
| name        |   Name of the customer (unique)                                        |
+-------------+------------------------------------------------------------------------+
| password    |   Password for registering the customer in the platform (optional)     |
+-------------+------------------------------------------------------------------------+
| fleet_type  |   Fleet type that the customer wants to use                            |
+-------------+------------------------------------------------------------------------+
| icon        |   Custom icon (in base64 format) to be used by the customer (optional) |
+-------------+------------------------------------------------------------------------+
| strategy    |   Custom strategy file in the format ``module.file.Class`` (optional)  |
+-------------+------------------------------------------------------------------------+
| delay       |   Agent's execution time start, in seconds  (optional)                 |
+-------------+------------------------------------------------------------------------+

A Taxi agent is defined by the following fields:

+---------------------------------------------------------------------------------------------+
|  Taxi                                                                                       |
+------------------+--------------------------------------------------------------------------+
|  Field           |  Description                                                             |
+==================+==========================================================================+
| class            |   Custom agent file in the format ``module.file.Class``                  |
+------------------+--------------------------------------------------------------------------+
| position         |   Initial coordinates of the transport (optional)                        |
+------------------+--------------------------------------------------------------------------+
| name             |   Name of the transport (unique)                                         |
+------------------+--------------------------------------------------------------------------+
| password         |   Password for registering the transport in the platform (optional)      |
+------------------+--------------------------------------------------------------------------+
| speed            |   Speed of the transport (in meters per second)  (optional)              |
+------------------+--------------------------------------------------------------------------+
| fleet_type       |   Fleet type that the customer wants to use                              |
+------------------+--------------------------------------------------------------------------+
| optional         |   **fleet**: The fleet manager's JID to be subscribed to (optional)      |
+------------------+--------------------------------------------------------------------------+
| icon             |   Custom icon (in base64 format) to be used by the transport  (optional) |
+------------------+--------------------------------------------------------------------------+
| strategy         |   Custom strategy file in the format ``module.file.Class`` (optional)    |
+------------------+--------------------------------------------------------------------------+
| delay            |   Agent's execution time start, in seconds  (optional)                   |
+------------------+--------------------------------------------------------------------------+

A FleetManager agent fields are defined as follows:

+--------------------------------------------------------------------------------------+
|  Fleet Manager                                                                       |
+-------------+------------------------------------------------------------------------+
|  Field      |  Description                                                           |
+=============+========================================================================+
| name        |   Name of the manager (unique)                                         |
+-------------+------------------------------------------------------------------------+
| password    |   Password for registering the manager in the platform (optional)      |
+-------------+------------------------------------------------------------------------+
| fleet_type  |   Fleet type that the agent manages                                    |
+-------------+------------------------------------------------------------------------+
| icon        |   Custom icon (in base64 format) to be used by the manager  (optional) |
+-------------+------------------------------------------------------------------------+
| strategy    |   Custom strategy file in the format ``module.file.Class``  (optional) |
+-------------+------------------------------------------------------------------------+

Finally, we show an example of a taxi service configuration file featuring four customers, two transports and a fleet manager.
This configuration file includes:

    * One taxi with a fixed initial position and another with a random initial position.
    * One customer with fixed origin and destination coordinates.
    * Three customers with random origin and destination coordinates.

.. code-block:: json

    {
    "fleets": [
        {
            "name": "fleet1",
            "password": "secret",
            "fleet_type": "taxi"
        }
    ],
    "transports": [
        {
            "class": "simfleet.common.lib.transports.models.taxi.TaxiAgent",
            "position": [
                39.470390,
                -0.356541
            ],
            "name": "taxi1",
            "password": "secret",
            "speed": 2000,
            "fleet_type": "taxi",
            "optional": {
                "fleet": "fleet1@localhost"
            },
            "icon": "taxi",
            "delay": 0
        },
        {
            "class": "simfleet.common.lib.transports.models.taxi.TaxiAgent",
            "name": "taxi2",
            "password": "secret",
            "speed": 2000,
            "fleet_type": "taxi",
            "optional": {
                "fleet": "fleet1@localhost"
            },
            "icon": "taxi"
        }
    ],
    "customers": [
        {
            "class": "simfleet.common.lib.customers.models.taxicustomer.TaxiCustomerAgent",
            "position": [
                39.45874369,
                -0.34011479
            ],
            "destination": [
                39.494655,
                -0.361639
            ],
            "name": "taxicustomer1",
            "password": "secret",
            "fleet_type": "taxi",
            "delay": 5
        },
        {
            "class": "simfleet.common.lib.customers.models.taxicustomer.TaxiCustomerAgent",
            "name": "taxicustomer2",
            "password": "secret",
            "fleet_type": "taxi",
            "delay": 5
        },
        {
            "class": "simfleet.common.lib.customers.models.taxicustomer.TaxiCustomerAgent",
            "name": "taxicustomer3",
            "password": "secret",
            "fleet_type": "taxi",
            "delay": 7
        },
        {
            "class": "simfleet.common.lib.customers.models.taxicustomer.TaxiCustomerAgent",
            "name": "taxicustomer4",
            "password": "secret",
            "fleet_type": "taxi",
            "delay": 10
        }
    ],
    "stations": [],
    "stops": [],
    "lines": [],
    "vehicles": [],
    "simulation_name": "taxis",
    "max_time": 100,
    "transport_strategy": "simfleet.common.lib.transports.strategies.taxi.FSMTaxiBehaviour",
    "customer_strategy": "simfleet.common.lib.customers.strategies.taxicustomer.AcceptFirstRequestBehaviour",
    "fleetmanager_strategy": "simfleet.common.lib.fleet.strategies.fleetmanager.DelegateRequestBehaviour",
    "fleetmanager_name": "fleetmanager",
    "fleetmanager_password": "fleetmanager_passwd",
    "host": "localhost",
    "http_port": 9000,
    "http_ip": "localhost"
    }

Electric taxi service simulation
--------------------------------

This transportation mode represents the same taxi service explained previously, with the modification that taxis are now
modeled as electric vehicles with a given autonomy level. A transport's autonomy will decrease as it serves customer requests.
The electric taxis check their autonomy level before each trip and may decide to recharge their batteries at a

HISTORY.rst
=======
History
=======

2.0.1 (2025-01-28)
------------------
* Minor bug fixes
* Updated documentation

2.0.0 (2025-01-23)
------------------
* New brand release
* Upgraded to SPADE 4
* Full code refactor
* New class hierarchy
* New documentation

1.0.3 (2022-12-13)
------------------
* Update third-party libs

1.0.2 (2021-02-14)
------------------
* Fixed bug with Vuex version

1.0.1 (2019-11-07)
------------------

* SPADE and pandas version upgraded.
* Stop simulation fixed.
* Aesthetic changes.
* Minor bug fixes.
* Updated documentation.

1.0.0 (2019-11-05)
------------------

* Moved from a taxi simulator to a generic fleet simulator.
* Updated documentation.
* Added support for different cities.
* Directory agent now sends all the info.
* Fixed bug of staying the corresponding time in the station when charging.
* Changed logger to loguru library.
* Removed fuel from transport popup, now is current_autonomy/max_autonomy.
* Concurrent charging in stations now allowed through TimeoutBehavior.
* Custom icons added.
* Removed agents introduction from GUI.
* Added specific parameters in scenario file (now config file).
* CLI simplified.
* Changed cli to config file.
* Control of free places and status for StationAgent.
* Fuel refill behavior between TransportAgent and StationAgent.
* Refactoring from passenger to Customer
* Refactoring from taxi to Transport
* Refactoring from coordinator to fleet manager

0.4.1 (2019-01-07)
------------------

* Fixed bug when checking if the simulation is finished.

0.4.0 (2018-10-25)
------------------

* Improved the concurrent creation of agents.
* Added stop and clear buttons to the interface.
* Added download button for getting results in excel and json formats.
* Documentation updated.

0.3.0 (2018-10-01)
------------------

* Migrated to SPADE 3.
* Documentation highly improved.
* Helper functions added and refined.
* Javascript framework included: VueJS
* Routes centralized with a Route agent.
* UI improved.

0.2 (2017-11-15)
------------------

* Added scenario loading feature.

0.1.3 (2017-11-15)
------------------

* Fixed minor bugs.

0.1.1 (2017-11-14)
------------------

* Added documentation.

0.1.0 (2017-11-03)
------------------

* First release on PyPI.



simfleet/config/settings.py
import json
from loguru import logger

from simfleet.utils.reflection import load_class
from simfleet.utils.helpers import get_bbox_from_location

def hide_passwords(item, key=None):
    if isinstance(item, dict):
        d = dict()
        for newk, newv in item.items():
            d[newk] = hide_passwords(newv, newk)
    elif isinstance(item, list):
        d = list()
        for i in item:
            d.append(hide_passwords(i))
    else:
        d = "*" * len(item) if isinstance(key, str) and "password" in key else item
    return d


class SimfleetConfig(object):
    """
    A scenario object reads a file with a JSON representation of a scenario and is used to create the participant agents.
    """

    def __init__(self, filename=None, name=None, max_time=None, verbose=None):
        """
        The SimfleetConfig constructor reads the JSON file and sets.
        Args:
            filename (str): the name of the scenario file
        """

        self.__config = dict()

        self.__config["fleets"] = []
        self.__config["transports"] = []
        self.__config["customers"] = []
        self.__config["stations"] = []
        self.__config["vehicles"] = []  # New vehicle
        # Bus line
        self.__config["stops"] = []
        self.__config["lines"] = []

        if filename:
            self.load_config(filename)

        self.__config["simulation_name"] = self.__config.get("simulation_name", name)
        self.__config["max_time"] = self.__config.get("max_time", max_time)
        self.__config["verbose"] = self.__config.get("verbose", verbose)
        self.__config["simulation_password"] = self.__config.get("simulation_password", "secret")

        #New coords
        if self.__config.get("coords"):
            if self.__config.get("zoom"):
                input_location = get_bbox_from_location(self.__config["coords"], self.__config["zoom"])
                logger.debug(
                    "BoundingBox for {} is {}".format(self.__config["coords"], input_location[1])
                )
                self.__config["coords"] = input_location
            else:
                logger.debug("Default value 12 for Zoom variable")
                input_location = get_bbox_from_location(self.__config["coords"], 12)
                self.__config["zoom"] = 12
                self.__config["coords"] = input_location
        else:
            #raise Exception("Could not find coordinates for the entered location")
            logger.debug(
                "Could not find coordinates for the entered location. Default coordinates: Valencia, ES"
            )
            default_location = get_bbox_from_location("Valencia, ES", 11.75)
            self.__config["zoom"] = self.__config.get("zoom", 11.75)
            self.__config["coords"] = default_location

        #self.__config["coords"] = self.__config.get("coords", [39.47, -0.37])
        #self.__config["zoom"] = self.__config.get("zoom", 12)

        self.__config["transport_strategy"] = self.__config.get(
            "transport_strategy", "simfleet.common.lib.transports.strategies.taxi.FSMTaxiBehaviour"
        )
        self.__config["customer_strategy"] = self.__config.get(
            "customer_strategy", "simfleet.common.lib.customers.strategies.taxicustomer.AcceptFirstRequestBehaviour"
        )
        self.__config["fleetmanager_strategy"] = self.__config.get(
            "fleetmanager_strategy", "simfleet.common.lib.fleet.strategies.fleetmanager.DelegateRequestBehaviour"
        )
        self.__config["directory_strategy"] = self.__config.get(
            "directory_strategy", "simfleet.common.agents.directory.DirectoryStrategyBehaviour"
        )
        self.__config["station_strategy"] = self.__config.get(
            "station_strategy", "simfleet.common.lib.stations.models.chargingstation.ChargingService"
        )
        #New vehicle
        self.__config["vehicle_strategy"] = self.__config.get(
            "vehicle_strategy", "simfleet.common.lib.vehicles.strategies.vehicle.FSMOneShotVehicleBehaviour"
        )
        # New statistics
        self.__config["mobility_metrics"] = self.__config.get(          #Metric - Renombrar
            "mobility_metrics", "simfleet.metrics.lib.mobilitystatistics.MobilityStatisticsClass"
        )

        # Bus line
        #self.__config["bus_stop_strategy"] = self.__config.get(
        #    "bus_stop_strategy", "simfleet.common.lib.stations.models.busstop.BusStopStrategyBehaviour")

        self.__config["fleetmanager_name"] = self.__config.get(
            "fleetmanager_name", "fleetmanager"
        )
        self.__config["fleetmanager_password"] = self.__config.get(
            "fleetmanager_passwd", "fleetmanager_passwd"
        )
        self.__config["route_host"] = self.__config.get(
            #"route_host", "http://router.project-osrm.org/"
            "route_host", "http://osrm.gti-ia.upv.es/"
        )
        self.__config["route_name"] = self.__config.get("route_name", "route")
        self.__config["route_password"] = self.__config.get(
            "route_passwd", "route_passwd"
        )
        self.__config["directory_name"] = self.__config.get(
            "directory_name", "directory"
        )
        self.__config["directory_password"] = self.__config.get(
            "directory_passwd", "directory_passwd"
        )

        self.__config["host"] = self.__config.get("host", "127.0.0.1")
        self.__config["xmpp_port"] = self.__config.get("xmpp_port", 5222)
        self.__config["http_port"] = self.__config.get("http_port", 9000)
        self.__config["http_ip"] = self.__config.get("http_ip", "127.0.0.1")

        logger.debug("Config loaded: {}".format(self))

    def load_config(self, filename):
        with open(filename, "r") as f:
            logger.info("Reading config {}".format(filename))
            self.__config.update(json.load(f))

    @property
    def num_managers(self):
        try:
            return len(self.__config["fleets"])
        except KeyError:
            return 0

    @property
    def num_transport(self):
        try:
            return len(self.__config["transports"])
        except KeyError:
            return 0

    @property
    def num_customers(self):
        try:
            return len(self.__config["customers"])
        except KeyError:
            return 0

    @property
    def num_stations(self):
        try:
            return len(self.__config["stations"])
        except KeyError:
            return 0

    #New vehicle
    @property
    def num_vehicles(self):
        try:
            return len(self.__config["vehicles"])
        except KeyError:
            return 0

    # Bus line
    @property
    def num_stops(self):
        try:
            return len(self.__config["stops"])
        except KeyError:
            return 0

    @property
    def num_lines(self):
        try:
            return len(self.__config["lines"])
        except KeyError:
            return 0

    def __getitem__(self, item):
        return self.__config[item]

    def __getattr__(self, item):
        if item != "__config" and item in self.__config:
            return self.__config[item]
        else:
            return super().__getattribute__(item)

    def __setattr__(self, key, value):
        if "__config" in self.__dict__ and key in self.__config:
            self.__config[key] = value
        else:
            super().__setattr__(key, value)

    def __str__(self):
        d = hide_passwords(self.__config)
        return json.dumps(d, indent=4)


def set_default_metrics(mobility_metrics):      #Renombrar
    class_dict = {}

    class_dict['mobility_metrics'] = load_class(mobility_metrics)     #Crear lista de metricas

    return class_dict

def set_default_strategies(
        directory_strategy,
        fleetmanager_strategy,
        transport_strategy,
        customer_strategy,
        station_strategy,
        vehicle_strategy,   #New vehicle
        #bus_stop_strategy,  #Bus line
):
    """
    Gets the strategy strings and loads their classes. This strategies are prepared to be injected into any
    new transport or customer agent.
    Args:
        directory_strategy (str): the path to the directory strategy
        fleetmanager_strategy (str): the path to the fleetmanager strategy
        transport_strategy (str): the path to the transport strategy
        customer_strategy (str): the path to the customer strategy
        station_strategy (str): the path to the station strategy
    """

    class_dict = {}

    class_dict['directory'] = load_class(directory_strategy)
    class_dict['fleetmanager'] = load_class(fleetmanager_strategy)
    class_dict['transport'] = load_class(transport_strategy)
    class_dict['customer'] = load_class(customer_strategy)
    class_dict['station'] = load_class(station_strategy)
    class_dict['vehicle'] = load_class(vehicle_strategy)  #New vehicle
    #class_dict['stop'] = load_class(bus_stop_strategy)  # Bus line

    logger.debug(
        "Loaded default strategy classes: {}, {}, {}, {} and {}".format(
            class_dict['directory'],
            class_dict['fleetmanager'],
            class_dict['transport'],
            class_dict['customer'],
            class_dict['station'],
            class_dict['vehicle'],  #New vehicle
            #class_dict['stop'],  # Bus line
        )
    )

    return class_dict
